File: php/src/StatementPrinter.php
<?php

declare(strict_types=1);

namespace Theatrical;

use Error;
use NumberFormatter;

class StatementPrinter
{
    /**
     * @param array<string, Play> $plays
     */
    public function print(Invoice $invoice, array $plays): string
    {
        $totalAmount = 0;
        $volumeCredits = 0;

        $result = "Statement for {$invoice->customer}\n";
        $format = new NumberFormatter('en_US', NumberFormatter::CURRENCY);

        foreach ($invoice->performances as $performance) {
            $play = $plays[$performance->playId];
            $performanceCost = $this->performancePlayCost($play, $performance);

            $volumeCredits = $this->getVolumeCredits($performance, $volumeCredits, $play);

            // print line for this order
            $result .= "  {$play->name}: {$format->formatCurrency($performanceCost / 100, 'USD')} ";
            $result .= "({$performance->audience} seats)\n";

            $totalAmount += $performanceCost;
        }

        $result .= "Amount owed is {$format ->formatCurrency($totalAmount / 100, 'USD')}\n";
        $result .= "You earned {$volumeCredits} credits";
        return $result;
    }

    private function tragedyAmount(Performance $performance): int|float
    {
        $thisAmount = 40000;
        if ($performance->audience > 30) {
            $thisAmount += 1000 * ($performance->audience - 30);
        }
        return $thisAmount;
    }

    private function comedyAmount(Performance $performance): int|float
    {
        $thisAmount = 30000;
        if ($performance->audience > 20) {
            $thisAmount += 10000 + 500 * ($performance->audience - 20);
        }
        $thisAmount += 300 * $performance->audience;
        return $thisAmount;
    }

    private function performancePlayCost(Play $play, Performance $performance): int|float
    {
        switch ($play->type) {
            case 'tragedy':
                return $this->tragedyAmount($performance);

            case 'comedy':
                return $this->comedyAmount($performance);

            default:
                throw new Error("Unknown type: {$play->type}");
        }
    }

    private function getVolumeCredits(Performance $performance, mixed $volumeCredits, Play $play): mixed
    {
// add volume credits
        $volumeCredits += max($performance->audience - 30, 0);
        // add extra credit for every ten comedy attendees
        if ($play->type === 'comedy') {
            $volumeCredits += floor($performance->audience / 5);
        }
        return $volumeCredits;
    }
}

File: dart/lib/statement_printer.dart
import 'dart:math';

import 'package:intl/intl.dart';
import 'package:theatrical_plays/invoice.dart';
import 'package:theatrical_plays/play.dart';

class StatementPrinter {

  String print(Invoice invoice, Map<String, Play> plays) {
    var totalAmount = 0;
    var volumeCredits = 0;
    var result = 'Statement for ${invoice.customer}\n';

    final frmt = NumberFormat.currency(locale: 'en_US', symbol: '\$');

    for (var perf in invoice.performances) {
      final play = plays[perf.playID];
      var thisAmount = 0;

      switch (play!.type) {
        case 'tragedy':
          thisAmount = 40000;
          if (perf.audience > 30) {
            thisAmount += 1000 * (perf.audience - 30);
          }
          break;
        case 'comedy':
          thisAmount = 30000;
          if (perf.audience > 20) {
            thisAmount += 10000 + 500 * (perf.audience - 20);
          }
          thisAmount += 300 * perf.audience;
          break;
        default:
          throw Exception('unknown type: ${play.type}');
      }

      // add volume credits
      volumeCredits += max(perf.audience - 30, 0);
      // add extra credit for every ten comedy attendees
      if ('comedy' == play.type) volumeCredits += (perf.audience ~/ 5).floor();

      // print line for this order
      result += '  ${play.name}: ${frmt.format(thisAmount ~/ 100)} (${perf.audience} seats)\n';
      totalAmount += thisAmount;
    }
    result += 'Amount owed is ${frmt.format(totalAmount ~/ 100)}\n';
    result += 'You earned $volumeCredits credits\n';
    return result;
  }
}